### PR TITLE
fix(integration): external approval is always created when initiating a ticket

### LINKED ISSUE #67: [Bug]: A new instance of external approval is always created when a ticket is initiated
### ODC version

4.2.0

### OB version

Regardless of OB version

### What happened?

When a ticket is initiated, a new instance of external approval is always created, even though the approval node list contains no external approval node.

### What did you expect to happen?

External approval instance should be created only when a external approval node is needed.

### How can we reproduce it (as minimally and precisely as possible)?

1. Config external approval integration;
2. Start a ticket in which the approval node list does not contain external approval;
3. Check the external approval system, you will find a new instance is created.

### Anything else we need to know?

_No response_

### Cloud

_No response_

### PR DESCRIPTION
**Pull request by smallsheeeep**
#### What this PR does / why we need it:
Fix bug: A new instance of external approval is always created when a ticket is initiated

#### Which issue(s) this PR fixes:
Fixes #67 
Close #67 

#### Special notes for your reviewer:
Already passed testing in BPMS.

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/flow/UserTaskInstanceRepository.java [hunk 1]
```diff
@@ -72,7 +72,8 @@ List<UserTaskInstanceEntity> findByResourceRoleIdentifierIn(
     @Transactional
     @Query("update UserTaskInstanceEntity as ut set ut.userTaskId=:#{#param.userTaskId},ut.status=:#{#param.status},"
             + "ut.approved=:#{#param.approved},ut.operatorId=:#{#param.operatorId},ut.comment=:#{#param.comment},"
-            + "ut.expireIntervalSeconds=:#{#param.expireIntervalSeconds} where ut.id=:#{#param.id}")
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/flow/UserTaskInstanceRepository.java:75]: Is it necessary to delete the update of `externalFlowInstanceId` here

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/instance/FlowInstanceConfigurer.java [hunk 2]
```diff
@@ -208,10 +211,28 @@ protected FlowInstanceConfigurer next(@NonNull FlowApprovalInstance nextNode,
             userTaskConsumer.accept(builder);
             return builder;
         });
-        targetExecution.next(userTaskBuilder);
+        if (Objects.nonNull(nextNode.getExternalApprovalId())) {
+            String serviceTaskName = "create_external_approval_task_" + nextNode.getId();
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/instance/FlowInstanceConfigurer.java:215]: Would it be better to keep the naming format consistent with others?
```Java
FlowNodeType.APPROVAL_TASK.name() + "_external_approval_task_" + nextNode.getId();
```

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/instance/FlowInstanceConfigurer.java [hunk 3]
```diff
@@ -208,10 +211,28 @@ protected FlowInstanceConfigurer next(@NonNull FlowApprovalInstance nextNode,
...
+            ServiceTaskBuilder serviceTaskBuilder = nullSafeGetNodeBuilder(serviceTaskName, nextNode,
+                    () -> new ServiceTaskBuilder(serviceTaskName, CreateExternalApprovalTask.class));
+            if (this.requiresActivityIdAndName) {
+                flowableAdaptor.setFlowableElement(nextNode, new FlowableElement(serviceTaskBuilder));
+            }
+            String gatewayName = "create_external_approval_gateway_" + nextNode.getId();
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/instance/FlowInstanceConfigurer.java:221]: same as above

**Comment by krihy** [comment]
Relational test case passed ?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    private ApprovalClient approvalClient;
+    private final RetryExecutor retryExecutor = RetryExecutor.builder().retryIntervalMillis(1000).retryTimes(3).build();
+
+    @Override
+    protected void run(DelegateExecution execution) throws Exception {
+        String activityId = execution.getCurrentActivityId();
+        String processDefinitionId = execution.getProcessDefinitionId();
+        Optional<Optional<Long>> flowInstanceIdOpt = retryExecutor.run(
+                () -> flowableAdaptor.getFlowInstanceIdByProcessDefinitionId(processDefinitionId), Optional::isPresent);
+        if (!flowInstanceIdOpt.isPresent() || !flowInstanceIdOpt.get().isPresent()) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java:60]: what will the flow be when this exception is thrown?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            throw new IllegalStateException(
+                    "Can not find flow instance id by process definition id " + processDefinitionId);
+        }
+        Long flowInstanceId = flowInstanceIdOpt.get().get();
+        Optional<FlowApprovalInstance> instanceOpt =
+                flowableAdaptor.getApprovalInstanceByActivityId(activityId, flowInstanceId);
+        if (!instanceOpt.isPresent()) {
+            log.warn("Flow approval instance does not exist, activityId={}, flowInstanceId={}", activityId,
+                    flowInstanceId);
+            throw new IllegalStateException("Can not find instance by activityId " + activityId);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java:72]: same as above

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        FlowApprovalInstance flowApprovalInstance = instanceOpt.get();
+        Long externalApprovalId = flowApprovalInstance.getExternalApprovalId();
+        if (Objects.nonNull(externalApprovalId)) {
+            log.info("Start create external approval instance, flowApprovalInstanceId={}, externalApprovalId={}",
+                    flowApprovalInstance.getId(), externalApprovalId);
+            IntegrationConfig config = integrationService.detailWithoutPermissionCheck(externalApprovalId);
+            ApprovalProperties properties = ApprovalProperties.from(config);
+            TemplateVariables variables = FlowTaskUtil.getTemplateVariables(execution.getVariables());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java:81]: don't put `TemplateVariables` into flow variables. `TemplateVariables` contains `Map<String, Object>`, which may lead to error~

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/flow/UserTaskInstanceRepository.java (hunk 1 above)
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/flow/UserTaskInstanceRepository.java:75]: Emmm, on the contrary, there is adding the update of `externalFlowInstanceId`. The pursose is to update external approval at flow runtime.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/instance/FlowInstanceConfigurer.java (hunk 2 above)
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/instance/FlowInstanceConfigurer.java:215]: get!

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/model/RuntimeTaskConstants.java [hunk 7]
```diff
@@ -34,5 +34,5 @@ public class RuntimeTaskConstants {
     public static final Integer DEFAULT_TASK_CHECK_INTERVAL_SECONDS = 5;
     public static final String RISKLEVEL_DESCRIBER = "riskLevelDescriber";
     public static final String RISKLEVEL = "riskLevel";
-    public static final String SQL_CHECK_RESULT = "sql_check_result";
```
[the remark this answers is not in the document]
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/model/RuntimeTaskConstants.java:37]: `SQL_CHECK_RESULT` is not used

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java (hunk 4 above)
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java:60]: The approval node will be terminated, but the flow instance must be terminated by user manually.
The status update is not as expected, I will continue solve this problem.

Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java:99]: New: If then flowApprovalInstance is not found, then just return and escape the ServiceTask.
If create external approval instance failed, then mark node execute failed and the flow is terminated.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java (hunk 6 above)
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/CreateExternalApprovalTask.java]: But, the `TemplateVariables` is needed by approval integration, and used to build HTTP body.
In preview testing, it works as expected. So can we keep it as before?

**Comment by smallsheeeep** [comment]
Change to new PR: https://github.com/oceanbase/odc/pull/140

**Pull request closed without merging**